### PR TITLE
Added CR to CRLF conversion support for ASCII downloads

### PR DESCRIPTION
**Pull request by andrew-shulgin**
As per https://github.com/giampaolo/pyftpdlib/pull/492#issuecomment-477135219

### CODE REVIEW

**Comment by andrew-shulgin** [comment]
Please tell if this acceptable, and I'll update CHANGES.rst if true.

**Comment by giampaolo** [comment]
Hm... this adds a lot of complexity. If os.sep is "\r" on OSX 9 only I would say it's better to leave this out.

**Comment by giampaolo** [comment]
What I would do is this:

```
if type == 'a' and os.linesep != '\r\n':
     assert os.sep == '\n', "platform not supported"
```

**Comment by giampaolo** [comment]
Here's what I would do instead:
```diff
--- a/pyftpdlib/handlers.py
+++ b/pyftpdlib/handlers.py
@@ -2835,8 +2835,12 @@ class FTPHandler(AsyncChat):
         """Set current type data type to binary/ascii"""
         type = line.upper().replace(' ', '')
         if type in ("A", "L7"):
-            self.respond("200 Type set to: ASCII.")
-            self._current_type = 'a'
+            if os.sep in ('\n', '\r\n'):
+                self.respond("200 Type set to: ASCII.")
+                self._current_type = 'a'
+            else:
+                self.respond('504 ASCII mode not supported on this platform')
+
         elif type in ("I", "L8"):
             self.respond("200 Type set to: Binary.")
             self._current_type = 'i'
```

**Comment by andrew-shulgin** [comment]
As I stated in the prevoius PR, the platform-native separator doesn't matter much, because any file can have CR line ending on any platform.

**Comment by andrew-shulgin** [comment]
According to [Wikipedia](https://en.wikipedia.org/wiki/Newline), CR line endings are native to:
- Commodore 8-bit machines (C64, C128)
- Acorn BBC
- ZX Spectrum
- TRS-80
- Apple II family
- Oberon
- classic Mac OS
- MIT Lisp Machine
- OS-9

Thus, any file created on these platforms has CR line endings.
To be honest, it seems that any of these platforms is widely used nowadays.

**Comment by giampaolo** [comment]
> As I stated in the prevoius PR, the platform-native separator doesn't matter much, because any file can have CR line ending on any platform, e.g. downloaded from the web.

But we cannot know that, nor make assumptions. It's RFC-959 which dictates to convert `os.sep` to `\r\n` when sending a file in ASCII mode.

Also, one question: you modified `FileProducer` (meaning RETR, aka file going from server to client). What about the other way around? Does the same problem affect `DTPHandler._posix_ascii_data_wrapper`?

**Comment by andrew-shulgin** [comment]
`DTPHandler._posix_ascii_data_wrapper` works correctly, because the client must always send CRLF-separated data if ASCII transfer mode is being used.

**Comment by andrew-shulgin** [comment]
>But we cannot know that

We can just detect and convert CR or LF line-endings on-the-fly; this exact behaviour is implemented in this PR.

**Comment by giampaolo** [comment]
> DTPHandler._posix_ascii_data_wrapper works correctly, because the client must always send CRLF-separated data if ASCII transfer mode is being used.

Oh you're right.

> We can just detect and convert CR or LF line endings on-the-fly; this exact behaviour is implemented in this PR.

Mm.. but is this really desirable? Couldn't "\r" be legitimate (aka it should not be converted)? It would be interesting to know what proftpd and vsftpd do in this regard.

**Comment by andrew-shulgin** [comment]
>It would be interesting to know what proftpd and vsftpd do in this regard.

I agree, going to check it soon.

**Comment by andrew-shulgin** [comment]
I checked that and conclude: both ProFTPd and vsftpd do not convert stand-alone CRs into CRLFs.

**Comment by lurch** [comment]
> CR line endings are native to:
ZX Spectrum

Oh noes, you mean I can't run pyftpdlib on my Speccy?! :scream_cat: :keyboard: :cold_sweat: :rofl:

File: pyftpdlib/handlers.py [hunk 1]
```diff
@@ -1031,29 +1032,42 @@ def __init__(self, file, type):
...
             self._data_wrapper = self._posix_ascii_data_wrapper
         else:
             self._data_wrapper = None
 
     def _posix_ascii_data_wrapper(self, chunk):
         """The data wrapper used for sending data in ASCII mode on
         systems using a single line terminator, handling those cases
-        where CRLF ('\r\n') gets delivered in two chunks.
+        where CRLF ('\r\n') gets delivered in two chunks as well as
+        cased where a file has mixed line endings.
```
Comment (lurch) [inline pyftpdlib/handlers.py]: cased -> cases

**Comment by lurch** [comment]
WRT the unit tests, I wonder if it might make sense to also have tests with `\r` or `\n` as the very first character of the string, to make sure any edge-cases get caught? :man_shrugging: 

Note that these review-comments don't imply that I "endorse" this change; that's entirely up to @giampaolo :slightly_smiling_face:

File: pyftpdlib/handlers.py (hunk 1 above)
Comment (andrew-shulgin) [reply pyftpdlib/handlers.py]: Fixed, thanks

**Comment by giampaolo** [comment]
> I checked that and can conclude: both ProFTPd and vsftpd do not convert stand-alone CRs into CRLFs. On Linux, at least.

OK, I'm gonna close this out then. Thanks.
